common/archiver/provider/provider.go

```diff
@@ -137,6 +139,22 @@ func (p *archiverProvider) GetHistoryArchiver(scheme, serviceName string) (archi
 			return nil, err
 		}
 
+		p.Lock()
+		defer p.Unlock()
+		if existingHistoryArchiver, ok := p.historyArchivers[archiverKey]; ok {
+			return existingHistoryArchiver, nil
+		}
+		p.historyArchivers[archiverKey] = historyArchiver
+		return historyArchiver, nil
```

Comment:
This is my fault, I should do the assignment outside the `switch` statement and return `ErrUnknownScheme` in the `default` case.

Can you help do this small refactoring?

Reply:
done!

Reply:
Sorry. Actually I mean the entire part can be moved out of the switch statement, not just the assignment on L147.

```
	var err error
	var historyArchiver archiver.HistoryArchiver
	switch scheme {
	case filestore.URIScheme:
		// assign file store implementation to historyArchiver, err
	case terrablob.URIScheme:
		// assign gcp implemention to historyArchiver, err
	default:
		return nil, ErrUnknownScheme
	}
	if err != nil {
		return nil, err
	}

	p.Lock()
	defer p.Unlock()
	if existingHistoryArchiver, ok := p.historyArchivers[archiverKey]; ok {
		return existingHistoryArchiver, nil
	}
	p.historyArchivers[archiverKey] = historyArchiver
	return historyArchiver, nil
```

Reply:
my fault!. Now there is no more duplicate code.